crates/runtime/src/table.rs

```diff
@@ -186,20 +310,48 @@ impl Table {
             return Err(Trap::wasm(ir::TrapCode::TableOutOfBounds));
         }
 
-        let srcs = src_index..src_index + len;
-        let dsts = dst_index..dst_index + len;
+        // Check if the source and destination are the same table
+        // This ensures we don't `borrow` and `borrow_mut` the same underlying RefCell
+        let same_table = ptr::eq(dst_table, src_table);
 
-        // Note on the unwraps: the bounds check above means that these will
-        // never panic.
-        //
-        // TODO(#983): investigate replacing this get/set loop with a `memcpy`.
-        if dst_index <= src_index {
-            for (s, d) in (srcs).zip(dsts) {
-                dst_table.set(d, src_table.get(s).unwrap()).unwrap();
-            }
-        } else {
-            for (s, d) in srcs.rev().zip(dsts.rev()) {
-                dst_table.set(d, src_table.get(s).unwrap()).unwrap();
+        let src_range = src_index as usize..src_index as usize + len as usize;
+        let dst_range = dst_index as usize..dst_index as usize + len as usize;
+
+        unsafe {
+            match dst_table.element_type() {
+                TableElementType::Func => dst_table.with_funcrefs_mut(|dst| {
+                    let dst = dst.unwrap();
```

Comment:
I'll fix.

Reply:
Fixed this in the `wasmtime` crate as returning a "table out of bounds" trap (the only trap code that would make sense to return from this runtime function) here for mismatched table types seems wrong; this is an API bug anyway since mismatched table types shouldn't pass module validation.